pkg/issuer/acme/prepare.go

```diff
@@ -41,287 +40,463 @@ const (
 		}
-	}()
+	}
+	crt.Status.ACMEStatus().Order.Challenges = newChallengeList
 
-	a.recorder.Eventf(crt, v1.EventTypeNormal, reasonPresentChallenge, messagePresentChallenge, challengeType, auth.domain)
-	err = solver.Present(ctx, crt, auth.domain, token, key)
+	err = utilerrors.NewAggregate(errs)
 	if err != nil {
-		return nil, fmt.Errorf("error presenting acme authorization for domain %q: %s", auth.domain, err.Error())
+		crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorPresentError, err.Error(), false)
+		return err
 	}
 
-	a.recorder.Eventf(crt, v1.EventTypeNormal, reasonSelfCheck, messageSelfCheck, auth.domain)
-	err = solver.Wait(ctx, crt, auth.domain, token, key)
+	return nil
+}
+
+// presentChallenge will process a challenge by talking to the acme server and
+// obtaining up to date status information.
+// If the challenge is still in a pending state, it will 'present' the
+// challenge using the appropriate solver.
+// It will then run the 'Check' function in order to determine whether the
+// challenge has propegated (e.g. to DNS resolvers, or to the ingress
+// controller).
+// If the check fails, an error will be returned.
+// Otherwise, it will return nil.
+func (a *Acme) presentChallenge(ctx context.Context, cl client.Interface, crt *v1alpha1.Certificate, ch v1alpha1.ACMEOrderChallenge) error {
+	acmeCh, err := cl.GetChallenge(ctx, ch.URL)
 	if err != nil {
-		return nil, fmt.Errorf("error waiting for key to be available for domain %q: %s", auth.domain, err.Error())
+		return err
 	}
 
-	challenge, err = cl.Accept(ctx, challenge)
-	if err != nil {
-		return nil, fmt.Errorf("error accepting acme challenge for domain %q: %s", auth.domain, err.Error())
+	switch acmeCh.Status {
+	case acme.StatusValid:
+		return nil
+	case acme.StatusInvalid, acme.StatusDeactivated, acme.StatusRevoked:
+		acmeErrReason := "unknown reason"
+		if acmeCh.Error != nil {
+			acmeErrReason = acmeCh.Error.Error()
+		}
+		return fmt.Errorf("challenge for domain %q failed: %s", ch.Domain, acmeErrReason)
+	case acme.StatusPending, acme.StatusProcessing:
+	default:
+		return fmt.Errorf("unknown acme challenge status %q for domain %q", acmeCh.Status, ch.Domain)
 	}
 
-	glog.V(4).Infof("waiting for authorization for domain %s (%s)...", auth.domain, challenge.URI)
-	authorization, err := cl.WaitAuthorization(ctx, challenge.URI)
+	solver, err := a.solverFor(ch.Type)
 	if err != nil {
-		return nil, fmt.Errorf("error waiting for authorization for domain %q: %s", auth.domain, err.Error())
+		return err
 	}
 
-	if authorization.Status != acme.StatusValid {
-		return nil, fmt.Errorf("expected acme domain authorization status for %q to be valid, but it is %q", auth.domain, authorization.Status)
+	err = solver.Present(ctx, crt, ch)
+	if err != nil {
+		return err
 	}
 
-	a.recorder.Eventf(crt, v1.EventTypeNormal, successObtainedAuthorization, messageObtainedAuthorization, auth.domain)
+	ok, err := solver.Check(ch)
```

Comment:
This check when done immediately after `solver.Present()` is very much likely to fail, making cert-manger to call `presentChallenge` again on  next attempts. But on subsequent attempts  it can never reach this line, because `solver.Present()` is not idempotent (at least with some providers) and fails with error, not giving solver.Check() a chance  to detect propagation and report success.

Interim fix for it is in https://github.com/munnerz/cert-manager/pull/7

Reply:
So it *is* expected that it will fail numerous times and keep retrying, however it should retry with an exponential backoff: 2 * (2^numFailures), e.g. 2s, 8s, 16s, 32s, 64s etc (with a cap of 1 minute iirc).

What commit hash did you build your binary from? You need to be using `43373cd` or later (preferably HEAD).

If you *are* seeing this issue despite being on the latest commit ref., I will need to investigate before merging as it will cause undue load on the ACME server.

Reply:
I can confirm, that backoff is not working  when presentChallenge() returns error, I didn't look into it as it doesn't bother me much :)